core/src/com/rallyhealth/weepickle/v1/core/ValidatingVisitor.scala

```diff
@@ -0,0 +1,243 @@
+package com.rallyhealth.weepickle.v1.core
+
+import java.time.Instant
+
+import scala.annotation.tailrec
+import scala.collection.mutable.ListBuffer
+import scala.util.control.NonFatal
+import scala.util.{Failure, Success, Try}
+
+/**
+  * A slight variation of JsonPointerVisitor, where we
+  * add a *sequence* of JSON Pointers to the result of the delegate Visitor
+  * rather than throwing a single exception for the first error encountered by the delegate.
+  *
+  * JSON Pointer is standardized by RFC 6901 and commonly used by JSON Schema.
+  *
+  * Useful for debugging failures. May add ~10-20% overhead depending on the parser, document size, number of errors, etc..
+  *
+  * In this microbenchmark parsing a 640 KB, 24 K line Json with two errors it added ~20% (i.e., about 1 ms):
+  *
+  * [info] BenchTest.play_All     thrpt   20   57.468 ± 1.367  ops/s
+  * [info] BenchTest.wee_1_First  thrpt   20  229.404 ± 5.225  ops/s
+  * [info] BenchTest.wee_2_All    thrpt   20  184.232 ± 3.377  ops/s
+  *
+  * (But still much faster than Play)
+  *
+  * @see https://tools.ietf.org/html/rfc6901
+  */
+object ValidatingVisitor {
+
+  type Errors = Seq[JsonPointerVisitor.JsonPointerException]
+
+  implicit class Ops[T, J](delegate: Visitor[T, J]) {
+
+    def returningAllErrors: Visitor[T, Either[Errors, J]] = new ValidatingVisitor(delegate, RootHasPath)
+  }
+
+  /**
+    * Internally, the paths form a linked list back to the root by the visitors themselves.
+    * Compared to something like a List[String] or List[Object], this does not require
+    * extra String allocation or boxing unless we actually ask for the path.
+    */
+  trait HasPath {
```

Comment:
Can `HasPath` and `HasMultiplePaths` be private?

Reply:
done